packages/@aws-cdk/aws-s3/test/auto-delete-objects.test.ts

```diff
@@ -0,0 +1,250 @@
+  expect(stack).not.toHaveResource('Custom::AutoDeleteObjects',
+    { BucketName: { Ref: 'MyBucketTwoC7437026' } });
+});
+
+test('iam policy is created', () => {
+  const stack = new cdk.Stack();
+
+  new s3.Bucket(stack, 'MyBucket', {
+    removalPolicy: cdk.RemovalPolicy.DESTROY,
+    autoDeleteObjects: true,
+  });
+
+  expect(stack).toHaveResource('AWS::IAM::Role', {
+    Policies: [
+      {
+        PolicyName: 'Inline',
+        PolicyDocument: {
+          Version: '2012-10-17',
+          Statement: [
+            {
+              Effect: 'Allow',
+              Resource: '*',
+              Action: [
+                's3:GetObject*',
+                's3:GetBucket*',
+                's3:List*',
+                's3:DeleteObject*',
+              ],
+            },
+          ],
+        },
+      },
+    ],
+  });
+});
+
+test('throws if autoDeleteObjects is enabled but if removalPolicy is not specified', () => {
+  const stack = new cdk.Stack();
+
+  expect(() => new s3.Bucket(stack, 'MyBucket', { autoDeleteObjects: true })).toThrowError(/removal policy/);
+});
+
+describe('custom resource handler', () => {
+
+  const mockS3Client = {
+    listObjectVersions: jest.fn().mockReturnThis(),
+    deleteObjects: jest.fn().mockReturnThis(),
+    promise: jest.fn(),
+  };
+
+  jest.mock('aws-sdk', () => {
+    return { S3: jest.fn(() => mockS3Client) };
+  });
+
+  beforeEach(() => {
+    mockS3Client.deleteObjects.mockClear();
+    mockS3Client.listObjectVersions.mockClear();
+    mockS3Client.promise.mockClear();
+  });
+
+  test('does nothing on create event', async () => {
+    const event: Partial<AWSLambda.CloudFormationCustomResourceCreateEvent> = {
+      RequestType: 'Create',
+      ResourceProperties: {
+        ServiceToken: 'Foo',
+        BucketName: 'MyBucket',
+      },
+    };
+    await invokeHandler(event);
+
+    expect(mockS3Client.listObjectVersions).not.toHaveBeenCalled();
+    expect(mockS3Client.deleteObjects).not.toHaveBeenCalled();
+  });
+
+  test('does nothing on update event', async () => {
+    const event: Partial<AWSLambda.CloudFormationCustomResourceUpdateEvent> = {
+      RequestType: 'Update',
+      ResourceProperties: {
+        ServiceToken: 'Foo',
+        BucketName: 'MyBucket',
+      },
+    };
+    await invokeHandler(event);
+
+    expect(mockS3Client.listObjectVersions).not.toHaveBeenCalled();
+    expect(mockS3Client.deleteObjects).not.toHaveBeenCalled();
+  });
+
+  test('deletes no objects on delete event when bucket has no objects', async () => {
+    mockS3Client.listObjectVersions().promise.mockResolvedValue({ Versions: [] }); // this counts as one call
+
+    const event: Partial<AWSLambda.CloudFormationCustomResourceDeleteEvent> = {
+      RequestType: 'Delete',
+      ResourceProperties: {
+        ServiceToken: 'Foo',
+        BucketName: 'MyBucket',
+      },
+    };
+    await invokeHandler(event);
+
+    expect(mockS3Client.listObjectVersions.mock.calls.length).toEqual(2);
```

Comment:
1. We should also assert it was called with the correct arguments (with `toHaveBeenCalledWith` assertion probably)
2. Have you tried to use `toHaveBeenCalledTimes` assertion instead of using `mock.calls`?

(similarly for other tests)

Reply:
1. Sure I'll try this. 2. Looks like I overlooked this matcher - thanks for catching. 👍